File: DS_and_Algos/Javascript/find_minimum_in_rotated_sorted_array.js
var findMin = function (nums) {
  let min = 0;
  let max = nums.length - 1;
  let mid;

  if (nums.length === 1 || nums[min] < nums[max]) {
    return nums[0];
  }

  while (min <= max) {
    let mid = Math.floor((min + max) / 2);

    if (nums[mid] < nums[mid - 1]) {
      return nums[mid];
    }

    if (nums[mid] > nums[max]) {
      min = mid + 1;
    } else {
      max = mid - 1;
    }
  }
  return mid;
};

console.log(findMin([1, 2]));

File: DS_and_Algos/Python/Algorithms/BinarySearch/find_minimum_in_rotated_sorted_array.py
class Solution:
    def findMin(self, nums: list[int]) -> int:
        # If the middle is equal to both the firs and the last, add 1 to left , subtract 1 from the right
        # Since there will be two independently sorted sub arrays, find the smallest one
        # If middle is greater than the right, go right
        # If the middle is less than the right, go left

        left, right = 0, len(nums) - 1
        
        while left <= right:
            mid = (left + right) // 2

            print(f"mid is {mid}, right is {right} and left is {left}")

            if nums[mid] == nums[left] and nums[mid] == nums[right]:
                left += 1
                right -= 1
            
            elif nums[left] <= nums[mid] and nums[left] <= nums[right]:
                left = mid - 1
            
            else:
                right = mid + 1

ExampleSolution = Solution()
print(ExampleSolution.findMin([3, 3, 1, 3]))